onnxmltools/utils/float16_converter.py

```diff
@@ -0,0 +1,113 @@
+            if hasattr(q, 'initializer'):
+                for n in q.initializer:
+                    if n.data_type == onnx_proto.TensorProto.FLOAT:
+                        n.data_type = onnx_proto.TensorProto.FLOAT16
+                        # convert float_data (float) to float16 and write to int32_data
+                        int_list = _npfloat16_to_int(np.float16(n.float_data))
+                        n.int32_data[:] = int_list
+                        n.float_data[:] = []
+                        # convert n.raw_data (bytes) to float
+                        float32_list = []
+                        for i in range(len(n.raw_data)//4):
+                            float32_list += unpack('f', n.raw_data[i*4:(i+1)*4])
+                        # convert float to float16
+                        int_list = _npfloat16_to_int(np.float16(float32_list))
+                        # convert float16 to bytes and write back to raw_data
+                        n.raw_data = b''
+                        for num in int_list:
+                            n.raw_data += num.to_bytes(2, byteorder='little', signed=False)
+            # graph.input (ValueInfoProto)
+            if hasattr(q, 'input'):
+                for n in q.input:
+                    # check if the input has type of ValueInfoProto (GraphProto) or string (NodeProto)
+                    if hasattr(n, 'type'):
+                         if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                            n.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+            # graph.output (ValueInfoProto)
+            if hasattr(q, 'output'):
+                for n in q.output:
+                    # check if the output has type of ValueInfoProto (GraphProto) or string (NodeProto)
+                    if hasattr(n, 'type'):
+                        if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                            n.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+            # if q is graph, process graph.value_info (ValueInfoProto)
+            if hasattr(q, 'value_info'):
+                for n in q.value_info:
+                    if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                        n.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+            # if q is node.attribute, process node.attribute.t and node.attribute.tensors (TensorProto)
+            if hasattr(q, 't'):
+                if q.t.data_type == onnx_proto.TensorProto.FLOAT:
+                    q.t.data_type = onnx_proto.TensorProto.FLOAT16
+                    int_list = _npfloat16_to_int(np.float16(q.t.float_data))
+                    q.t.int32_data[:] = int_list
+                    q.t.float_data[:] = []
+                    # convert q.t.raw_data (bytes) to float
+                    float32_list = []
```

Comment:
If `float_data` is set, the field `raw_data` must be empty so we should skip the subsequent lines. To determine which filed is set, please see [this](https://github.com/onnx/onnx/blob/4b4085c2e9d5a944651a2dd0dfdd20ef452bdcdf/onnx/onnx.in.proto#L319).